---
title: "BRCA1 first look"
author: "Karsten Bach"
date: '`r Sys.Date()`'
output:
  html_notebook:
    toc: true
    toc_depth: 2
    toc_float: true
    theme: united
    highlight: tango
    code_folding: hide
---
***

# Quality and Sanity checks
```{r, message=FALSE,warning=FALSE}
library(scran)
library(scater)
library(plyr)
library(dplyr)
library(knitr)
library(ggplot2)
library(Rtsne)
library(cowplot)
library(Matrix)
library(gridExtra)
source("functions.R")

# Load Data
dataList <- readRDS("../data/Robjects/ExpressionList_QC.rds")
sfs <- read.csv("../data/Robjects/SizeFactors.csv")
m <- dataList[["counts"]]
pD <- dataList[["phenoData"]]
pD <- left_join(pD,sfs)
fD <- dataList[["featureData"]]
rm(dataList)
```

## Overview over samples
We prepared a total of 14 scRNA-seq libraries from two 10x Chips (7 per chip), resulting in a total of `r ncol(m)` cells.
CBLT represents the wildtype control animal.
WKBR61.4dT is the tumor and WKBR61.4dM is the remaining glands pooled from the same animal.

***
__I mostly use the age covariate to split the different conditions, note that the 53 weeks old animal is WT!__

```{r, message=FALSE}
sumry <- group_by(pD, Condition) %>%
    summarize("# Before QC"=n(),
	      "Total molecules"=median(UmiSums),
	      "Genes Detected"=median(GenesDetected),
	      "Age"=unique(Age))
sumryAQC <- filter(pD, PassAll) %>%
    group_by(Condition) %>%
    summarize("# After QC"=n())
sumry <- full_join(sumry,sumryAQC)
sumry <- sumry[,c(1,5,2,6,4,3)]
kable(sumry)
```

## Most of the variation in cell number is between biological replicates
The samples from each animal were split into two technical replicates that were loaded onto two different chips.
We can thus compare whether the variation in number of cells is greater between biological or technical replicates.
It appears that the number of cells varies mainly between biological replicates, indicating that the cell quantification before loading mostly determines the number of cells sequenced rather than differences in capture efficiency.
```{r}
fp <- group_by(pD[!pD$AnimalHadTumor,], SampleID) %>%
    summarize(Number=n())
fp$Condition <- as.factor(substring(fp$SampleID,1,nchar(fp$SampleID)-1))
ggplot(fp, aes(x=Condition, y=Number)) +
    geom_point()
```

## Comparison to FACS Analysis
A fraction of cells was taken from the sample that was submitted to sequencing was analysed using flow cytometry.
I compared the fraction of EpCam+ cells between FACS and scRNA-seq data to see if there's any bias in capturing efficiency between epithelial vs. non-epithelial.
### No bias for epithelial cells
```{r, message=FALSE, warning=FALSE}
facs <- read.csv("../data/misc/FACS_data.csv")
colnames(facs)[1] <- "Condition"
anno_full <- read.csv("../data/Robjects/Cluster_all.csv")
pD.sub <- left_join(anno_full,pD)
summ <- group_by(pD.sub, Condition) %>% 
    summarize(scRNA_Epithelial=(sum((Class=="Epithelial"))/n())*100,
	      scRNA_Basal=(sum((Cluster=="C14"))/sum(Class=="Epithelial"))*100,
	      scRNA_Luminal=(sum((Cluster %in% c("C20","C12")))/sum(Class=="Epithelial"))*100)
	      
summ <- left_join(summ,facs)
ggplot(summ, aes(x=scRNA_Epithelial, y=Epithelial, color=Condition)) +
    geom_point() +
    geom_smooth(method="lm",aes(color=NULL)) +
    geom_abline(intercept=0,slope=1,lty="dashed",color="grey50") +
    xlab("% Epithelial from scRNA-seq Data") +
    ylab("% Epithelial from FACS")
```

### Bias for capturing luminal cells vs basal cells
According to the FACS data the fraction of luminal cells is much smaller than it is the scRNA-seq data. Either the definition of "luminal" or "basal" from the FACS and scRNA-seq don't fully overlap or basal cells are less likely to be captured.
**I need to go back and double check the FACS data.**
```{r, message=FALSE, warning=FALSE}
summ <- summ[!grepl("61.4",summ$Condition),]
ggplot(summ, aes(x=scRNA_Luminal, y=luminal, color=Condition)) +
    geom_point() +
    geom_smooth(method="lm",aes(color=NULL)) +
    geom_abline(intercept=0,slope=1,lty="dashed",color="grey50") +
    xlab("% Luminal from scRNA-seq Data") +
    ylab("% Luminal from FACS") +
    xlim(0,100) +
    ylim(0,100)
```


## Quality of sequencing data
Overall the quality looks good, we do have less genes detected and less UMI counts per cell on average.
My feeling is that this is because we get less molecules from non-epithelial cells, whether this is because of less mRNA or different lysis efficiency I don't know.
```{r, fig.width=12, fig.height=12}
filt.gd1 <- min(pD$GenesDetected[pD$PassAll])
filt.gd2 <- max(pD$GenesDetected[pD$PassAll])
filt.lb1 <- min(pD$UmiSums[pD$PassAll])
filt.lb2 <- max(pD$UmiSums[pD$PassAll])
filt.mt <- max(pD$prcntMito[pD$PassAll])

# Quality Control
# Sequencing Depth and Genes detected
genesDetected <- ggplot(pD, aes(x=SampleID,y=GenesDetected,fill=Condition)) +
    geom_violin(draw_quantiles=0.5)+
    scale_y_log10() +
    ylab("Total number of genes detected") +
    xlab("") +
    geom_hline(yintercept=filt.gd1,lty="dashed",color="red") +
    geom_hline(yintercept=filt.gd2,lty="dashed",color="red") +
    scale_fill_brewer(palette="Dark2") +
    theme_bw() +
    theme(legend.position="none") +
    coord_flip()

LibrarySize <- ggplot(pD, aes(x=SampleID,y=UmiSums,fill=Condition)) +
    geom_violin(draw_quantiles=0.5)+
    scale_y_log10() +
    ylab("Total number of molecules") +
    xlab("") +
    geom_hline(yintercept=filt.lb1,lty="dashed",color="red") +
    geom_hline(yintercept=filt.lb2,lty="dashed",color="red") +
    scale_fill_brewer(palette="Dark2") +
    theme_bw() +
    theme(legend.position="none") +
    coord_flip()

cellViability <- ggplot(pD, aes(x=SampleID,y=prcntMito,fill=Condition)) +
    geom_violin(draw_quantiles=0.5)+
    ylab("Percentage of mitochondrial molecules") +
    xlab("") +
    geom_hline(yintercept=filt.mt,lty="dashed",color="red") +
    scale_fill_brewer(palette="Dark2") +
    theme_bw() +
    coord_flip()


mytheme <- gridExtra::ttheme_default(
    core = list(fg_params=list(cex = 0.8)),
    colhead = list(fg_params=list(cex = 0.8)),
    rowhead = list(fg_params=list(cex = 0.8)))

p1 <- tableGrob(sumry,rows=NULL,theme=mytheme)
p2 <- plot_grid(genesDetected, LibrarySize, cellViability,nrow=1,rel_widths=c(1,1,1.3))
p3 <- plot_grid(p2,p1,nrow=2)
print(p3)
```

```{r}
cells <- pD$barcode[pD$PassAll==TRUE]
m <- m[,cells]
pD <- pD[pD$barcode %in% cells,]
genes <- rowMeans(m)>0.01
m <- m[genes,]
```

## Batch effect is minor
In order to assess the batch effect between the two 10x chips, we can look at the technical replicates that were spread across the two chips.
Each tSNE represents cells from one animal, and is colored by the two replicates, that were loaded on two separate chips.
```{r, fig.width=10, fig.height=13, warning=FALSE, message=FALSE}
conds <- unique(pD$Condition[pD$AnimalHadTumor==FALSE])
pltlist <- list()
for (cond in conds) {
    tsn <- read.csv(paste0("../data/Robjects/tSNE_",cond,"_raw.csv"))
    pD.sub <- right_join(pD,tsn)
    pD.sub <- arrange(pD.sub, prcntMito)
    pltlist[[cond]] <- ggplot(pD.sub, aes(x=tSNE1, y=tSNE2, color=SampleID)) +
	geom_point() +
	theme_classic() +
	ggtitle(cond)
}

plot_grid(plotlist=pltlist,ncol=2)
```

# Cell type Inference 
After having compared cell type inference with or without tumor, we have decided to infer cell types with the tumor sample included as this yielded satisfactory results and will facilitate comparisons later one between tumor and non-tumor samples.
## tSNEs
```{r, fig.width=12, fig.height=12, warning=FALSE, message=FALSE}
tsn <- read.csv("../data/Robjects/tSNE_38-41-46-53_raw.csv")
anno_full <- read.csv("../data/Robjects/Cluster_all.csv")[,-1]
pD.sub <- left_join(pD,tsn)
pD.sub <- left_join(pD.sub,anno_full)

ggplot(pD.sub, aes(x=tSNE1, y=tSNE2, color=Cluster)) +
    geom_point() +
    ggtitle("Cluster") 

ggplot(pD.sub, aes(x=tSNE1, y=tSNE2, color=Class)) +
    geom_point() +
    ggtitle("Lineages")
```

## UMAP
```{r, fig.width=12, fig.height=12, warning=FALSE, message=FALSE}
umap <- read.csv("../data/Robjects/UMAP_all.csv")
pD.sub <- left_join(pD.sub,umap)
ggplot(pD.sub, aes(x=UMAP1, y=UMAP2, color=Cluster)) +
    geom_point() +
    ggtitle("Cluster") 
# facet_wrap(~Cluster_Full)

ggplot(pD.sub, aes(x=UMAP1, y=UMAP2, color=Class)) +
    geom_point() +
    ggtitle("Lineages")
```
